dubbo-xds/src/main/java/org/apache/dubbo/xds/security/authz/resolver/ConnectionCredentialResolver.java

```diff
@@ -0,0 +1,284 @@
+
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.extension.Activate;
+import org.apache.dubbo.common.logger.ErrorTypeAwareLogger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+import org.apache.dubbo.common.utils.NetUtils;
+import org.apache.dubbo.remoting.api.ChannelContextListener;
+import org.apache.dubbo.rpc.Invocation;
+import org.apache.dubbo.rpc.RpcContext;
+import org.apache.dubbo.rpc.RpcContextAttachment;
+import org.apache.dubbo.xds.security.authz.RequestCredential;
+import org.apache.dubbo.xds.security.authz.rule.RequestAuthProperty;
+
+import javax.net.ssl.SSLSession;
+
+import java.net.InetSocketAddress;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.grpc.netty.shaded.io.netty.channel.ChannelHandlerContext;
+import io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler;
+
+@Activate(order = -20)
+public class ConnectionCredentialResolver implements CredentialResolver, ChannelContextListener {
+
+    private final Map<String, ConnectionCredential> connectionInfos = new ConcurrentHashMap<>();
+
+    private static final ErrorTypeAwareLogger logger =
+            LoggerFactory.getErrorTypeAwareLogger(ConnectionCredentialResolver.class);
+
+    @Override
+    public void appendRequestCredential(URL url, Invocation invocation, RequestCredential requestCredential) {
+        requestCredential.add(
+                RequestAuthProperty.TARGET_VERSION,
+                invocation.getInvoker().getUrl().getVersion());
+        RpcContextAttachment serverContext = RpcContext.getServerContext();
+        requestCredential.add(
+                RequestAuthProperty.DIRECT_REMOTE_IP,
+                serverContext.getRemoteAddress().getHostName());
+        requestCredential.add(RequestAuthProperty.REMOTE_PORT, serverContext.getRemotePort());
+        requestCredential.add(RequestAuthProperty.REMOTE_APPLICATION, serverContext.getRemoteApplicationName());
+        requestCredential.add(RequestAuthProperty.REMOTE_GROUP, serverContext.getGroup());
+        requestCredential.add(RequestAuthProperty.DESTINATION_IP, serverContext.getLocalHost());
+        requestCredential.add(RequestAuthProperty.DESTINATION_PORT, serverContext.getLocalPort());
+
+        ConnectionCredential credential = connectionInfos.get(url.getIp() + ":" + url.getPort());
+        if (credential != null) {
+            requestCredential.add(RequestAuthProperty.CONNECTION_CREDENTIAL, credential);
+            requestCredential.add(RequestAuthProperty.REQUESTED_SERVER_NAME, credential.getSni());
+        }
+    }
+
+    @Override
+    public void onConnect(Object channelContext) {
+        if (channelContext instanceof ChannelHandlerContext) {
```

Comment:
Should we use the original ChannelHandlerContext instead of the grpc shaded version?

Reply:
Yes, it's fixed now